pkg/kubelet/dockershim/docker_container_cleanup_manager.go

```diff
@@ -0,0 +1,226 @@
+	if cleanupInfo == nil {
+		return
+	}
+
+	cm.mutex.Lock()
+	defer cm.mutex.Unlock()
+
+	if _, present := cm.pendingCleanupInfos.Get(containerID); present {
+		// shouldn't happen, as we only insert at container creation time - but to err on the side of
+		// caution, let's delete from and re-insert into the ordered map to ensure that the order
+		// stays chronological
+		klog.Errorf("duplicate cleanup info for container ID %q", containerID)
+		cm.pendingCleanupInfos.Delete(containerID)
+	}
+
+	cm.pendingCleanupInfos.Set(containerID, &containerCleanupInfoWithTimestamp{
+		cleanupInfo: cleanupInfo,
+		timestamp:   currentNanoTimestampFunc(),
+	})
+
+	if withFailureCount, present := cm.failedCleanupInfos[containerID]; present {
+		// this shouldn't happen either for the same reason; but same as above, if
+		// it does happen let's log it and maintain a sane internal state
+		klog.Errorf("duplicate cleanup info for container ID %q - had already tried to clean it up %v times unsuccessfully",
+			containerID, withFailureCount.failureCount)
+		delete(cm.failedCleanupInfos, containerID)
+	}
+}
+
+// performCleanup cleans up the given containerID.
+func (cm *containerCleanupManager) performCleanup(containerID string) {
+	cm.mutex.Lock()
+	defer cm.mutex.Unlock()
+
+	cm.unsafePerformCleanup(containerID)
+}
+
+// unsafePerformCleanup is the same as performCleanup, but assumes the manager's lock has already
+// been acquired.
+func (cm *containerCleanupManager) unsafePerformCleanup(containerID string) {
+	if cleanupInfoWithTimestamp, present := cm.pendingCleanupInfos.Delete(containerID); present {
+		cleanupInfo := cleanupInfoWithTimestamp.(*containerCleanupInfoWithTimestamp).cleanupInfo
+
+		if errors := cm.cleanupService.performContainerCleanup(containerID, cleanupInfo); len(errors) != 0 {
+			cm.failedCleanupInfos[containerID] = &containerCleanupInfoWithFailureCount{
+				cleanupInfo:  cleanupInfo,
+				failureCount: 1,
+			}
+		}
+	} else {
+		cm.unsafeRetryFailedCleanup(containerID)
+	}
+}
+
+// unsafeRetryFailedCleanup re-tries running a clean up that has already failed earlier. It assumes
+// the manager's lock has already been acquired.
+func (cm *containerCleanupManager) unsafeRetryFailedCleanup(containerID string) {
+	if cleanupInfoWithFailureCount, present := cm.failedCleanupInfos[containerID]; present {
+		if errors := cm.cleanupService.performContainerCleanup(containerID, cleanupInfoWithFailureCount.cleanupInfo); len(errors) == 0 {
+			delete(cm.failedCleanupInfos, containerID)
+		} else {
+			cleanupInfoWithFailureCount.failureCount++
+			if cleanupInfoWithFailureCount.failureCount >= maxFailures {
+				klog.Errorf("unable to clean up container %q, giving up after %v failures", containerID, cleanupInfoWithFailureCount.failureCount)
+				delete(cm.failedCleanupInfos, containerID)
+			}
+		}
+	}
+}
+
+// retryFailedCleanups retries running clean ups that have failed before.
+func (cm *containerCleanupManager) retryFailedCleanups() []string {
```

Comment:
nit - it might be good to prefix this with unsafe so its consistent with the other functions that assume the caller holds the mutex

Reply:
Good idea, done, thanks :)